### ISSUE TITLE
Android Platform multi-process may have a problem

### ISSUE THREAD

**Issue by yihome** [Tencent/MMKV#400]
I'm learning about MMKV source code ,in the InterProcessLock.cpp  i saw the code like this
```
bool FileLock::doLock(LockType lockType, bool wait) {
    if (!isFileLockValid()) {
        return false;
    }
    bool unLockFirstIfNeeded = false;

    if (lockType == SharedLockType) {
        m_sharedLockCount++;
        // don't want shared-lock to break any existing locks
        if (m_sharedLockCount > 1 || m_exclusiveLockCount > 0) {
            return true;
        }
    } else {
        m_exclusiveLockCount++;
        // don't want exclusive-lock to break existing exclusive-locks
        if (m_exclusiveLockCount > 1) {
            return true;
        }
        // prevent deadlock
        if (m_sharedLockCount > 0) {
            unLockFirstIfNeeded = true;
        }
    }
    return platformLock(lockType, wait, unLockFirstIfNeeded);
}
```
then if Process A has the ` LOCK_EX` type lock , if Process B want a `LOCK_SH` lock, the Process B will block in the `platformLock` method, but the m_sharedLockCount will increase，in the same time if another thread in Process B also want the  `LOCK_SH` lock, it will success in the code ` if (m_sharedLockCount > 1 || m_exclusiveLockCount > 0)` 
I'm not sure about this ,or you have some other mechanism to prevent this。

**Comment by lingol** [comment]
You're probably right. I will check it out ASAP.

**Comment by lingol** [comment]
I have checked the issue. Turns out MMKV works fine in such use cases. Here's why: when process B requires an exclusive inter-process and waiting for process A to finish, any subsequent call in any thread of process B will require an exclusive lock.
In a word, the use case you mention just doesn't exist in MMKV.

However, the design of `FileLock::doLock()` does have the fault, I will make it much more robust in the next version.
